samza-core/src/main/java/org/apache/samza/runtime/ClusterBasedProcessorLifecycleListener.java

```diff
@@ -0,0 +1,133 @@
+  private final Runnable onSignal;
+  private Thread shutdownHookThread = null;
+  private CountDownLatch shutdownLatch = new CountDownLatch(1);
+  private TaskConfig taskConfig;
+  private ProcessorLifecycleListener processorLifecycleListener;
+  private volatile Throwable containerException = null;
+
+  ClusterBasedProcessorLifecycleListener(Config config, ProcessorLifecycleListener processorLifecycleListener,
+      Runnable onSignal) {
+    this.taskConfig = new TaskConfig(config);
+    this.processorLifecycleListener = processorLifecycleListener;
+    this.onSignal = onSignal;
+  }
+
+  @Override
+  public void beforeStart() {
+    log.info("Before starting the container.");
+    addShutdownHook();
+    processorLifecycleListener.beforeStart();
+  }
+
+  @Override
+  public void afterStart() {
+    log.info("Container Started");
+    processorLifecycleListener.afterStart();
+  }
+
+  @Override
+  public void afterStop() {
+    log.info("Container Stopped");
+    processorLifecycleListener.afterStop();
+    removeShutdownHook();
+    shutdownLatch.countDown();
+  }
+
+  @Override
+  public void afterFailure(Throwable t) {
+    log.info("Container Failed");
+    containerException = t;
+    processorLifecycleListener.afterFailure(t);
+    removeShutdownHook();
+    shutdownLatch.countDown();
+  }
+
+  Throwable getContainerException() {
+    return containerException;
+  }
+
+  private void removeShutdownHook() {
+    try {
+      if (shutdownHookThread != null) {
+        removeJVMShutdownHook();
+      }
+    } catch (IllegalStateException e) {
+      // Thrown when then JVM is already shutting down, so safe to ignore.
+    }
+  }
+
+  private void addShutdownHook() {
+    shutdownHookThread = new Thread("Samza Container Shutdown Hook Thread") {
+      @Override
+      public void run() {
+        long shutdownMs = taskConfig.getShutdownMs();
+        log.info("Attempting to shutdown container from inside shutdownHook, will wait up to {} ms.", shutdownMs);
+        try {
+          onSignal.run();
+          boolean hasShutdown = shutdownLatch.await(shutdownMs, TimeUnit.MILLISECONDS);
+          if (hasShutdown) {
+            log.info("Shutdown complete");
+          } else {
+            log.error("Did not shut down within {} ms, exiting.", shutdownMs);
+            ThreadUtil.logThreadDump("Thread dump from Samza Container Shutdown Hook.");
+          }
+        } catch (InterruptedException e) {
+          log.error("Shutdown hook inturrupted while waiting on runLoop to shutdown");
+          ThreadUtil.logThreadDump("Thread dump from Samza Container Shutdown Hook.");
+        }
+      }
+    };
+    addJVMShutdownHook();
+  }
+
+  @VisibleForTesting
+  Thread getShutdownHookThread() {
+    return this.shutdownHookThread;
+  }
+
+  @VisibleForTesting
+  void addJVMShutdownHook() {
+    Runtime.getRuntime().addShutdownHook(shutdownHookThread);
```

Comment:
Since `shutdownHookThread` is a mutable instance variable, it makes it a little easier to accidentally call these before `shutdownHookThread` is actually set to a value (you currently have it right, but someone make change this code in the future).
Consider passing the `shutdownHookThread` as an argument to this method instead of using the instance var directly. Then the calling method is more aware that `shutdownHookThread` needs to be non-null.
Or you could check for null in this method.

Reply:
Makes sense. Fixed.